### PR TITLE
Simulation example of laser melting with recoil pressure

### PR DESCRIPTION
**Pull request by mschreter**
This PR introduces a first benchmark example to analyze the surface tension-recoil pressure interaction during selective laser melting to study deep keyhole dynamics.

The following assumptions are made:

- The temperature field is provided by an analytical relation for a constant moving laser heat source, following the approach by Mirkoohi et al. (2019): Heat source modeling in selective laser melting.
- Liquid (melted) regions with a temperature value smaller than the melting temperature are assumed to be solid, simply by setting the pressure and the velocity value of the overlapping fluid domain to zero.
- Marangoni forces are neglected.

As an example, the following simulation results can be obtained:

image

### CODE REVIEW

**Comment by peterrum** [comment]
The failing simplex test is related to https://github.com/dealii/dealii/issues/11204. Let me think about how to fix this tomorrow!

File: include/meltpooldg/melt_pool/melt_pool_operation.hpp [hunk 1]
```diff
@@ -135,6 +142,104 @@ namespace MeltPoolDG
...
+                    for (unsigned int v = 0; v < VectorizedArray<double>::size(); ++v)
+                      {
+                        alpha[v] = T[v] < T0 ? alpha0 : alpha0 - d_alpha0 * (T[v] - T0);
+                        AssertThrow(
+                          alpha[v] >= 0.0,
+                          ExcMessage(
+                            "The surface tension coefficient tends to be negative in "
+                            "some regions. Check the value of the temperature dependent surface "
+                            "tension coefficient."));
+                      }
```
Comment (peterrum) [inline include/meltpooldg/melt_pool/melt_pool_operation.hpp:227]: Do you want this check also in release mode (i.e. production run)? If not, I would use `Assert`.

File: include/meltpooldg/melt_pool/melt_pool_operation.hpp [hunk 2]
```diff
@@ -135,6 +142,104 @@ namespace MeltPoolDG
...
+                    const auto grad_T = temperature_val.get_gradient(q_index);
+
+                    Tensor<1, dim, VectorizedArray<double>> temp_surf_ten;
+
+                    for (unsigned int v = 0; v < VectorizedArray<double>::size(); ++v)
+                      for (unsigned int i = 0; i < dim; ++i)
+                        for (unsigned int j = 0; j < dim; ++j)
+                          temp_surf_ten[i][v] =
+                            (i == j) ? -(1. - n[i][v] * n[j][v]) * d_alpha0 * grad_T[j][v] :
+                                       (n[i][v] * n[j][v]) * d_alpha0 * grad_T[j][v];
```
Comment (peterrum) [inline include/meltpooldg/melt_pool/melt_pool_operation.hpp:215]: You don't need to loop here over the lanes. It should work if you replace `1` by `make_vectorized_array<double>(1)` or you write `- n[i] * n[j] + 1` (we definitely need to define some mixed operations within `base/vectorization.h`).

File: include/meltpooldg/melt_pool/melt_pool_operation.hpp [hunk 3]
```diff
@@ -135,6 +142,104 @@ namespace MeltPoolDG
...
+
+                    VectorizedArray<double> alpha;
+                    for (unsigned int v = 0; v < VectorizedArray<double>::size(); ++v)
+                      {
+                        alpha[v] = T[v] < T0 ? alpha0 : alpha0 - d_alpha0 * (T[v] - T0);
```
Comment (peterrum) [inline include/meltpooldg/melt_pool/melt_pool_operation.hpp:220]: This statment can also be written without the loops with https://www.dealii.org/developer/doxygen/deal.II/vectorization_8h.html#ac274b257ee3ff9666b2062b8ba56db8f.

File: include/meltpooldg/melt_pool/melt_pool_operation.hpp [hunk 4]
```diff
@@ -167,16 +273,60 @@ namespace MeltPoolDG
...
               for (unsigned int i = 0; i < dofs_per_cell; ++i)
-                temperature[local_dof_indices[i]] =
-                  analytical_temperature_field(support_points[local_dof_indices[i]],
-                                               level_set_as_heaviside[local_dof_indices[i]]);
+                {
+                  temperature[local_dof_indices[i]] =
+                    analytical_temperature_field(support_points[local_dof_indices[i]],
+                                                 level_set_as_heaviside[local_dof_indices[i]]);
+                  solid[local_dof_indices[i]] =
+                    is_solid_region(support_points[local_dof_indices[i]]) ? 1.0 : 0.0;
```
Comment (peterrum) [inline include/meltpooldg/melt_pool/melt_pool_operation.hpp:281]:
```suggestion
                  solid[local_dof_indices[i]] = is_solid_region(support_points[local_dof_indices[i]]);
```
... if you don't want implicit conversion, you could also introduce a static cast (what we do in deal.II).

File: include/meltpooldg/melt_pool/melt_pool_operation.hpp [hunk 5]
```diff
@@ -167,16 +273,60 @@ namespace MeltPoolDG
...
+                                flow_dof_handler.get_fe(),
+                                scratch_data->get_quadrature(flow_quad_idx),
+                                update_values);
+
+        const unsigned int dofs_per_cell = flow_dof_handler.get_fe().n_dofs_per_cell();
+
+        std::vector<types::global_dof_index> local_dof_indices(dofs_per_cell);
+
+        std::map<types::global_dof_index, Point<dim>> support_points;
+        DoFTools::map_dofs_to_support_points(MappingQGeneric<dim>(3),
```
Comment (peterrum) [inline include/meltpooldg/melt_pool/melt_pool_operation.hpp:304]: Is there a reason for 3?

**Review by peterrum — APPROVED** [review]
Looks good to me! I have left some comment but most belong in the category "midterm".

I am fine to merge even if the tests are not passing (indentation should however^^). There is already an open PR related to the simplex issue; the other tests are passing locally (on you machine and my machine) - I can investigate this issue at a later time (I have the feeling that it is related to the hardware).

File: include/meltpooldg/simulations/recoil_pressure/recoil_pressure.json [hunk 6]
```diff
@@ -2,9 +2,9 @@
   "base" : {
     "problem name": "melt_pool",
     "application name": "recoil_pressure",
-    "global refinements": "4",
+    "global refinements": "5",
```
[the remark this answers is not in the document]
Comment (mschreter) [reply include/meltpooldg/simulations/recoil_pressure/recoil_pressure.json:5]: That is a very good idea. I will put it on my list of issues.

**Pull request merged**